sympy/polys/domains/domain.py

```diff
@@ -1277,7 +1277,49 @@ def log(self, a, b):
         raise NotImplementedError
 
     def sqrt(self, a):
-        """Returns square root of ``a``. """
+        """Returns a (possibly inexact) square root of ``a``.
+
+        Explanation
+        ===========
+        There is no universal definition of "inexact square root" for all
+        domains. It is not recommended to implement this method for domains
+        other then :ref:`ZZ`.
+
+        See also
+        ========
+        exsqrt
+        """
+        raise NotImplementedError
+
+    def is_square(self, a):
+        """Returns whether ``a`` is a square in the domain.
+
+        Explanation
+        ===========
+        Returns ``True`` if there is an element ``b`` in the domain such that
+        ``b * b == a``, otherwise returns ``False``.
+
+        See also
+        ========
+        exsqrt
+        """
+        raise NotImplementedError
+
+    def exsqrt(self, a):
+        """Returns an exact square root of ``a``, or ``None`` if it doesn't
+        exist.
```

Comment:
I suppose this is not exact in the case of RR or CC.

Perhaps I would say
```
Principal square root of a within the domain if a is square.
```
The explanation can clarify what "principal" means.